include/wil/cppwinrt_authoring.h

```diff
@@ -111,6 +111,126 @@ struct single_threaded_rw_property : single_threaded_property<T>
 
 #endif // __WIL_CPPWINRT_AUTHORING_PROPERTIES_INCLUDED
 
+#if (!defined(__WIL_CPPWINRT_AUTHORING_INCLUDED_ICLASSFACTORY) && defined(__IClassFactory_INTERFACE_DEFINED__)) || \
+    defined(WIL_DOXYGEN) // class factory
+/// @cond
+#define __WIL_CPPWINRT_AUTHORING_INCLUDED_ICLASSFACTORY
+/// @endcond
+
+template <typename T, typename... Rest>
+struct class_factory : winrt::implements<class_factory<T, Rest...>, IClassFactory, winrt::no_weak_ref, Rest...>
+{
+    HRESULT __stdcall CreateInstance(IUnknown* outer, GUID const& iid, void** result) noexcept final
+    try
+    {
+        *result = nullptr;
+
+        if (!outer)
+        {
+            return winrt::make_self<T>().as(iid, result);
+        }
+        else
+        {
+            return CLASS_E_NOAGGREGATION;
+        }
+    }
+    CATCH_RETURN()
+
+    HRESULT __stdcall LockServer(BOOL lock) noexcept final
+    try
+    {
+        if (lock)
+        {
+            ++winrt::get_module_lock();
+        }
+        else
+        {
+            --winrt::get_module_lock();
+        }
+
+        return S_OK;
+    }
+    CATCH_RETURN()
+};
+
+#endif // !defined(__WIL_CPPWINRT_AUTHORING_INCLUDED_ICLASSFACTORY) && defined(__IClassFactory_INTERFACE_DEFINED__)
+
+#if (!defined(__WIL_CPPWINRT_AUTHORING_INCLUDED_COM_SERVER) && defined(__WIL_CPPWINRT_AUTHORING_INCLUDED_ICLASSFACTORY) && defined(_COMBASEAPI_H_)) || \
+    defined(WIL_DOXYGEN) // COM server
+/// @cond
+#define __WIL_CPPWINRT_AUTHORING_INCLUDED_COM_SERVER
+/// @endcond
+
+template <typename T>
+WI_NODISCARD_REASON("The class is unregistered when the returned value is destructed")
+unique_com_class_object_cookie
+    register_com_server(GUID const& guid, DWORD context = CLSCTX_LOCAL_SERVER, DWORD flags = REGCLS_MULTIPLEUSE)
+{
+    unique_com_class_object_cookie registration;
+    winrt::check_hresult(CoRegisterClassObject(
+        guid, winrt::make<class_factory<T, winrt::no_module_lock>>().get(), context, flags, registration.put()));
+    return registration;
+}
+
+template <typename... Types>
+struct clsid_array
+{
+    std::array<GUID, sizeof...(Types)> value;
+};
+
+template <typename... Ts>
+WI_NODISCARD_REASON("The classes are unregistered when the returned value is destructed")
+std::vector<unique_com_class_object_cookie> register_com_server(
+    wil::clsid_array<Ts...> const& clsids, DWORD context = CLSCTX_LOCAL_SERVER, DWORD flags = REGCLS_MULTIPLEUSE)

```

Comment:
I would recommend keeping the std::array parameter and making clsid_array a new overload.

The overload can simply pass it on to the std::array version

Reply:
I don't mind that, but let's see what @dunhor vote for.

Reply:
Gotcha, I see.

Reply:
Note that keeping the original `std::array` signature produces `error C2668: 'wil::register_com_server': ambiguous call to overloaded function`. I'd solve it by making `clsid_array`'s ctor explicit. I pushed a commit implementing it.

Btw, removing the original array signature on the other hand, allow both initializer-list style aka `{ _uuidof() }` and `make_array_of_x` to work. So it's literally the best of both world with less code, no explicit ctor etc. I guess keeping the original array signature makes the code "easier" to read?

Reply:
Hm, if the original style still works, then I retract my suggestion. Clever.

Reply:
The double braces are required for `std::array`. We would've needed `std::initializer_list` for single braces.

Reply:
`std::array` supports aggregate initialization, so single braces should be fine https://godbolt.org/z/o5cPPjYKx